[receiver/jmx] Add metric gatherer to artifacts

crobert-1 opened the pull request:
To properly support the jmx receiver in Docker, the dockerfile has to download the JMX metrics gatherer, and have it available to the collector. This change adds this functionality to the linux/macos dockerfile, as well as the windows dockerfile.

Also, adds an integration test getting metrics from a docker container.

jeffreyc-splunk commented on internal/buildscripts/packaging/fpm/common.sh:
```diff
@@ -65,6 +66,17 @@ create_user_group() {
+    mkdir -p "$buildroot/opt"
+
+    echo "Downloading ${JMX_METRIC_GATHERER_RELEASE_DL_URL}"
+    curl -sL "$JMX_METRIC_GATHERER_RELEASE_DL_URL" -o "$buildroot/opt/opentelemetry-java-contrib-jmx-metrics.jar"
```
Can you include the jar in `setup_files_and_permissions` below to ensure it is installed with the correct ownership/permissions?

crobert-1 replied:
Good idea, I've added it.

rmfitzpatrick commented on cmd/otelcol/Dockerfile:
```diff
@@ -53,6 +60,7 @@ COPY --from=otelcol --chown=999 /migratecheckpoint /
 COPY --from=otelcol --chown=999 /otel /etc/otel
 COPY --from=otelcol --chown=999 /otel/collector /etc/otel/collector
 COPY --from=agent-bundle --chown=999 /usr/lib/splunk-otel-collector /usr/lib/splunk-otel-collector
+COPY --from=jmx --chown=999 /tmp/opentelemetry-java-contrib-jmx-metrics.jar /opt/opentelemetry-java-contrib-jmx-metrics.jar
```
This is the [default path](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/jmxreceiver#jar_path-default-optopentelemetry-java-contrib-jmx-metricsjar) for the receiver so users would have to otherwise set it manually. Do any issues come to mind or is there an existing reason to mount /opt that could interfere?

jeffreyc-splunk replied:
Ah, makes sense.

jeffreyc-splunk commented on internal/buildscripts/packaging/msi/splunk-otel-collector.wxs:
```diff
@@ -67,8 +67,11 @@
+      <CustomAction Id="CopyJMXJAR" ExeCommand="xcopy /y &quot;[INSTALLDIR]opentelemetry-java-contrib-jmx-metrics.jar&quot; &quot;[WindowsVolume]\opt\opentelemetry-java-contrib-jmx-metrics.jar*&quot;" Directory="INSTALLDIR" Impersonate="no" Execute="deferred" Return="check" />
       <InstallExecuteSequence>
          <Custom Action="CopyConfig" After="InstallFiles">NOT CONFIG_FILE_EXISTS AND NOT Installed</Custom>
+         <Custom Action="CopyJMXJAR" After="InstallFiles">UPGRADINGPRODUCTCODE OR NOT Installed</Custom>
```
Would it be better to install the jar file directly to `\opt`?  Otherwise, I think we should add a check, similar to https://github.com/signalfx/splunk-otel-collector/blob/main/internal/buildscripts/packaging/msi/splunk-otel-collector.wxs#L58, to copy the jar file if it does not exist.

crobert-1 replied:
I've changed it to be installed automatically to `/opt`, in `./msi-build/build.sh`. I've removed this copy command.

jeffreyc-splunk commented on cmd/otelcol/Dockerfile.windows:
```diff
@@ -27,6 +27,22 @@ RUN Add-Type -AssemblyName System.IO.Compression.FileSystem ;\
+RUN Invoke-WebRequest -Uri "https://repo1.maven.org/maven2/io/opentelemetry/contrib/opentelemetry-jmx-metrics/$env:JMX_METRIC_GATHERER_RELEASE/opentelemetry-jmx-metrics-$env:JMX_METRIC_GATHERER_RELEASE.jar" -Outfile "c:\opt\opentelemetry-java-contrib-jmx-metrics.jar"
+
+ARG JAVA_VERSION="11.0.13_8"
+ARG OPENJDK_BASE_URL="https://github.com/AdoptOpenJDK/openjdk11-upstream-binaries/releases/download"
```
The AdoptOpenJDK project is archived.  Should we change it to https://github.com/adoptium/temurin11-binaries/releases, which is what we package in the [linux image and packages](https://github.com/signalfx/splunk-otel-collector/blob/main/internal/signalfx-agent/bundle/Dockerfile#L36)?

Also, if we want to automate the version update and ensure consistency with the linux version, we can try including this in https://github.com/signalfx/splunk-otel-collector/blob/main/.gitlab/update-openjdk.sh#L28.

crobert-1 replied:
Updated, thanks!

crobert-1 commented on cmd/otelcol/Dockerfile.windows:
```diff
@@ -27,6 +27,22 @@ RUN Add-Type -AssemblyName System.IO.Compression.FileSystem ;\
+# Download JMX Metric Gatherer
+ARG JMX_METRIC_GATHERER_RELEASE
+RUN New-Item -Path "C:\\" -Name "opt" -ItemType "directory"
+RUN Invoke-WebRequest -Uri "https://repo1.maven.org/maven2/io/opentelemetry/contrib/opentelemetry-jmx-metrics/$env:JMX_METRIC_GATHERER_RELEASE/opentelemetry-jmx-metrics-$env:JMX_METRIC_GATHERER_RELEASE.jar" -Outfile "c:\opt\opentelemetry-java-contrib-jmx-metrics.jar"
```
I'll update.

jeffreyc-splunk commented on cmd/otelcol/Dockerfile.windows:
```diff
@@ -27,6 +27,22 @@ RUN Add-Type -AssemblyName System.IO.Compression.FileSystem ;\
+# Download JMX Metric Gatherer
+ARG JMX_METRIC_GATHERER_RELEASE
+RUN New-Item -Path "C:\\" -Name "opt" -ItemType "directory"
+RUN Invoke-WebRequest -Uri "https://github.com/open-telemetry/opentelemetry-java-contrib/releases/download/v$env:JMX_METRIC_GATHERER_RELEASE/opentelemetry-jmx-metrics.jar" -Outfile "c:\opt\opentelemetry-java-contrib-jmx-metrics.jar"
```
```suggestion
RUN Invoke-WebRequest -Uri "https://github.com/open-telemetry/opentelemetry-java-contrib/releases/download/v${env:JMX_METRIC_GATHERER_RELEASE}/opentelemetry-jmx-metrics.jar" -Outfile "c:\opt\opentelemetry-java-contrib-jmx-metrics.jar"
```

crobert-1 commented on cmd/otelcol/Dockerfile:
```diff
@@ -35,6 +35,12 @@ RUN if [[ "$ARCH" = "amd64" || "$ARCH" = "arm64" ]]; then \
+FROM alpine:3.17.0 AS jmx
+ARG JMX_METRIC_GATHERER_RELEASE
+RUN JMX_METRICS_JAR=opentelemetry-jmx-metrics-${JMX_METRIC_GATHERER_RELEASE}.jar && \
+URL=https://repo1.maven.org/maven2/io/opentelemetry/contrib/opentelemetry-jmx-metrics/${JMX_METRIC_GATHERER_RELEASE}/$JMX_METRICS_JAR && \
```
Good idea, I've updated.

jeffreyc-splunk commented on cmd/otelcol/Dockerfile:
```diff
@@ -35,6 +35,12 @@ RUN if [[ "$ARCH" = "amd64" || "$ARCH" = "arm64" ]]; then \
 
+FROM alpine:3.17.0 AS jmx
+ARG JMX_METRIC_GATHERER_RELEASE
+RUN JMX_METRICS_JAR=opentelemetry-jmx-metrics.jar && \
```
Should the jar name match the [default path](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/jmxreceiver#jar_path-default-optopentelemetry-java-contrib-jmx-metricsjar), i.e. `opentelemetry-java-contrib-jmx-metrics.jar`, or will it be overridden?

crobert-1 replied:
You're right, it should match the default path. I renamed the paths to match the file being downloaded, forgetting the actual receiver's requirement here. Thanks!

jeffreyc-splunk commented on cmd/otelcol/Dockerfile.windows:
```diff
@@ -27,6 +27,22 @@ RUN Add-Type -AssemblyName System.IO.Compression.FileSystem ;\
+RUN Set-Variable -Name "ENCODED_VER" -Value "\"$env:JDK_VERSION\".replace(\"_\", \"%2B\")"; \
+    Invoke-WebRequest -Uri "${env:OPENJDK_BASE_URL}/jdk-${ENCODED_VER}/OpenJDK11U-jdk_x64_windows_hotspot_${env:JDK_VERSION}.zip" -OutFile "openjdk.zip" ; \
+    Expand-Archive "openjdk.zip" -DestinationPath "C:\java" ;
+RUN setx PATH "$env:JAVA_HOME;$env:PATH"
```
Not sure if this will set the env var correctly.  Maybe something like the following will work better:
```suggestion
RUN powershell.exe -Command $path = "${env:JAVA_HOME}\bin" + $env:path; Set-ItemProperty -Path 'HKLM:\SYSTEM\CurrentControlSet\Control\Session Manager\Environment\' -Name Path -Value $path
```
See https://github.com/MicrosoftDocs/Virtualization-Documentation/blob/main/windows-container-samples/golang/Dockerfile#L15.

crobert-1 replied:
It looks like we're using the servercore images, so I ended up following the [syntax there instead.](https://github.com/MicrosoftDocs/Virtualization-Documentation/blob/main/windows-container-samples/golang/Dockerfile.WindowsServerCore)

jeffreyc-splunk commented:
> The integration test is dependent on an official collector release containing the changes made in this PR. The integration test is passing locally, but will fail in github CI until a new release of the Splunk distribution of the OpenTelemetry Collector is created.
> 
> For this reason, I'm going to remove the integration test, and once this PR is merged, submit a new PR with the integration test.

The [integration test job](https://github.com/signalfx/splunk-otel-collector/blob/main/.github/workflows/integration-test.yml#L30) does build the `otelcol:latest` image for the branch.  Should the test be updated to use this local image instead?

crobert-1 commented:
> > The integration test is dependent on an official collector release containing the changes made in this PR. The integration test is passing locally, but will fail in github CI until a new release of the Splunk distribution of the OpenTelemetry Collector is created.
> > For this reason, I'm going to remove the integration test, and once this PR is merged, submit a new PR with the integration test.
> 
> The [integration test job](https://github.com/signalfx/splunk-otel-collector/blob/main/.github/workflows/integration-test.yml#L30) does build the `otelcol:latest` image for the branch. Should the test be updated to use this local image instead?

Ah, you're right, the [line here](https://github.com/signalfx/splunk-otel-collector/blob/main/.github/workflows/integration-test.yml#L127C1-L128C1) does what I needed. I'll add the test back in. I only saw the makefile was referring to quay's latest image and got turned around.

jeffreyc-splunk commented on cmd/otelcol/Dockerfile.windows:
```diff
@@ -27,6 +27,23 @@ RUN Add-Type -AssemblyName System.IO.Compression.FileSystem ;\
+
+ARG JDK_VERSION="11.0.19_7"
+ARG OPENJDK_BASE_URL="https://github.com/adoptium/temurin11-binaries/releases/download"
+ENV JAVA_HOME="C:\java\jdk-11.0.19+7"
```
```suggestion
ENV JAVA_HOME="C:\java\jdk-11.0.20+8"
```
whenever `JDK_VERSION` is updated.

May also be useful to dynamically set this path based on `JDK_VERSION`, or just rename the extracted jdk directory to something like `C:\java\jdk`.

crobert-1 replied:
Updated, and I'll try to re-use the version since the path is a sub-directory of the downloaded JDK, not the top level path.

jeffreyc-splunk commented on cmd/otelcol/Dockerfile.windows:
```diff
@@ -27,6 +27,23 @@ RUN Add-Type -AssemblyName System.IO.Compression.FileSystem ;\
+RUN New-Item -Path "C:\\" -Name "opt" -ItemType "directory"
+RUN Invoke-WebRequest -Uri "https://github.com/open-telemetry/opentelemetry-java-contrib/releases/download/v${env:JMX_METRIC_GATHERER_RELEASE}/opentelemetry-jmx-metrics.jar" -Outfile "c:\opt\opentelemetry-java-contrib-jmx-metrics.jar"
+
+ARG JDK_VERSION="11.0.19_7"
```
```suggestion
ARG JDK_VERSION="11.0.20_8"
```
Please rebase. The JDK was just updated to [`11.0.20_8`](https://github.com/signalfx/splunk-otel-collector/pull/3396) in the linux bundle.

crobert-1 replied:
Updated and rebased, thanks!

jeffreyc-splunk commented on cmd/otelcol/Dockerfile.windows:
```diff
@@ -27,6 +27,23 @@ RUN Add-Type -AssemblyName System.IO.Compression.FileSystem ;\
+
+ARG JDK_VERSION="11.0.20_8"
+ARG OPENJDK_BASE_URL="https://github.com/adoptium/temurin11-binaries/releases/download"
+ENV JAVA_HOME="C:\java\jdk-$JDK_VERSION.replace("_", "+")"
```
Not sure if `ENV` can handle the replace.  May need to do this in a `RUN` command with powershell.

jeffreyc-splunk commented on cmd/otelcol/Dockerfile.windows:
```diff
@@ -27,6 +27,26 @@ RUN Add-Type -AssemblyName System.IO.Compression.FileSystem ;\
+
+ARG JDK_VERSION="11.0.20_8"
+ARG OPENJDK_BASE_URL="https://github.com/adoptium/temurin11-binaries/releases/download"
+ENV JAVA_HOME="C:\java\jdk-$JDK_VERSION"
+RUN [Environment]::SetEnvironmentVariable('JAVA_HOME', "${Env:JAVA_HOME}.replace('_', '+')")
```
It may be simplest/safest to just use a set path, something like:
```suggestion
ENV JAVA_HOME="C:\java\jdk"
```
And then after the zip is downloaded and extracted, just rename the extracted directory:
```
RUN Move-Item C:\java\jdk-* C:\java\jdk
```

crobert-1 replied:
Sounds good to me. Looks like `Move-Item` [does](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.management/move-item?view=powershell-7.3) support regex, so I'll try this.

jeffreyc-splunk commented on .github/workflows/win-package-test.yml:
```diff
@@ -197,6 +197,7 @@ jobs:
           & ${{ github.workspace }}\.github\workflows\scripts\win-test-support-bundle.ps1 -mode "${{ matrix.MODE }}" -with_fluentd "${{ matrix.WITH_FLUENTD }}"
           Test-Path -Path "$env:ProgramFiles\Splunk\OpenTelemetry Collector\agent-bundle\python\python.exe"
           Test-Path -Path "$env:ProgramFiles\Splunk\OpenTelemetry Collector\agent-bundle\collectd-python"
+          Test-Path -Path "$env:opt\opentelemetry-java-contrib-jmx-metrics.jar"
```
Looks like `Test-Path` only returns true or false, and doesn't actually fail the test: https://github.com/signalfx/splunk-otel-collector/actions/runs/5695063950/job/15437772112#step:4:1451

```suggestion
          Resolve-Path -Path "$env:ProgramFiles\Splunk\OpenTelemetry Collector\agent-bundle\python\python.exe"
          Resolve-Path -Path "$env:ProgramFiles\Splunk\OpenTelemetry Collector\agent-bundle\collectd-python"
          Resolve-Path -Path "$env:SYSTEMDRIVE\opt\opentelemetry-java-contrib-jmx-metrics.jar"
```

jeffreyc-splunk commented on cmd/otelcol/Dockerfile:
```diff
@@ -53,6 +60,10 @@ COPY --from=otelcol --chown=999 /migratecheckpoint /
 COPY --from=agent-bundle --chown=999 /usr/lib/splunk-otel-collector /usr/lib/splunk-otel-collector
+COPY --from=jmx --chown=999 /tmp/opentelemetry-jmx-metrics.jar /opt/opentelemetry-java-contrib-jmx-metrics.jar
+
+#COPY --from=otelcol --chown=999 /tmp /tmp
+#ENV TMPDIR=/tmp
```
Let's remove this if it is no longer needed.
```suggestion
```

crobert-1 replied:
Good call, forgot it there!

jeffreyc-splunk commented on internal/buildscripts/packaging/msi/splunk-otel-collector.wxs:
```diff
@@ -67,6 +67,7 @@
 
       <!-- Copy the default agent config file to ProgramData if it does not already exist -->
       <CustomAction Id="CopyConfig" ExeCommand="xcopy /y &quot;[INSTALLDIR]agent_config.yaml&quot; &quot;[CommonAppDataFolder]Splunk\OpenTelemetry Collector\agent_config.yaml*&quot;" Directory="INSTALLDIR" Impersonate="no" Execute="deferred" Return="check" />
+      <!-- Copy the JMX Metric Gatherer JAR to opt if it does not already exist -->
```
```suggestion
```
Also, you may need to include `opt` in `<Directory>` above to package and install the jar since it's not getting copied anymore from `Program Files`.

crobert-1 replied:
I'm not following. Can you explain what you mean by including `opt` in `<Directory>` above? I understand it's not getting copied properly, but I'm not seeing how to fix that any way more simply than what was being done originally:
```
 <CustomAction Id="CopyJMXJAR" ExeCommand="xcopy /y &quot;[INSTALLDIR]opentelemetry-java-contrib-jmx-metrics.jar&quot; &quot;[WindowsVolume]\opt\opentelemetry-java-contrib-jmx-metrics.jar*&quot;" Directory="INSTALLDIR" Impersonate="no" Execute="deferred" Return="check" />
```

Should I take that solution and modify it for the current paths involved?

jeffreyc-splunk replied:
When we package/install the other files in `\Program Files`, we define the source and target paths in https://github.com/signalfx/splunk-otel-collector/blob/main/internal/buildscripts/packaging/msi/splunk-otel-collector.wxs#L14.  Since the jar is not currently included, I think we'll have to set the source path in https://github.com/signalfx/splunk-otel-collector/blob/main/internal/buildscripts/packaging/msi/make.ps1#L62 add a new `Directory` element in the wxs file.

jeffreyc-splunk replied:
> When we package/install the other files in `\Program Files`, we define the source and target paths in https://github.com/signalfx/splunk-otel-collector/blob/main/internal/buildscripts/packaging/msi/splunk-otel-collector.wxs#L14. Since the jar is not currently included, I think we'll have to set the source path in ~~https://github.com/signalfx/splunk-otel-collector/blob/main/internal/buildscripts/packaging/msi/make.ps1#L62~~ add a new `Directory` element in the wxs file.

Sorry, I referenced the file, it should be https://github.com/signalfx/splunk-otel-collector/blob/main/internal/buildscripts/packaging/msi/msi-builder/build.sh#L166

jeffreyc-splunk commented on cmd/otelcol/Dockerfile:
```diff
@@ -45,6 +51,7 @@ FROM scratch
 ENV SPLUNK_BUNDLE_DIR=/usr/lib/splunk-otel-collector/agent-bundle
 ENV SPLUNK_COLLECTD_DIR=${SPLUNK_BUNDLE_DIR}/run/collectd
 ENV PATH=${PATH}:${SPLUNK_BUNDLE_DIR}/bin
+ENV PATH=${PATH}:${SPLUNK_BUNDLE_DIR}/jre/bin
```
I wonder if adding `ENV JAVA_HOME=${SPLUNK_BUNDLE_DIR}/jre` helps with anything?

crobert-1 replied:
I checked and it's not set by default. If everything is currently working as expected I don't know if it's necessary, but also don't think it'd hurt. Do you think there's any specific benefit of setting it?

jeffreyc-splunk replied:
Looks like the it gets set for the smart agent (https://github.com/signalfx/splunk-otel-collector/blob/main/internal/signalfx-agent/pkg/core/config/config.go#L210), but wasn't sure if the jmx receiver needed it as well.  I guess if everything is working as expected without it, let's not fix things that aren't broken.

jeffreyc-splunk commented on tests/receivers/jmxreceiver/jmxreceiver_test.go:
```diff
@@ -0,0 +1,58 @@
+)
+
+func TestJMXReceiverProvidesAllJVMMetrics(t *testing.T) {
+	// JMX metrics are missing when running on arm so the test only checks a subset
+	// of what would be received otherwise.
```
Is this documented anywhere that we should include in our docs?

crobert-1 replied:
It's not documented anywhere that I've found. The only jmx collector [documentation from Splunk](https://docs.splunk.com/Observability/gdi/monitors-languages/jmx.html) that I've seen is pretty out of date, so there isn't a good spot to update there either. It's also not documented in the contrib repo, nor the java jmx metrics gatherer repo.

rmfitzpatrick replied:
In other instances where this is the case I have a feeling it's related to qemu and not an actual issue w/ arm. Once we have real arm instances in CI my assumption can be validated: https://github.com/signalfx/splunk-otel-collector/blob/f6dc932544b394f09635685afb57d66ec0069d1a/tests/receivers/discovery/host_observer_test.go#L28

jeffreyc-splunk commented:
Are any docs required?  At a minimum, I think we should document that the packages/images will install/include the jar.

jeffreyc-splunk commented:
Also, please add an entry to https://github.com/signalfx/splunk-otel-collector/blob/main/CHANGELOG.md.

rmfitzpatrick commented on CHANGELOG.md:
```diff
@@ -2,6 +2,10 @@
 
+### 💡 Enhancements 💡
+
+- (Splunk) `jmxreceiver`: Enable running in Docker on Windows and Linux ([#3262](https://github.com/signalfx/splunk-otel-collector/pull/3262))
```
```suggestion
- (Splunk) `jmxreceiver`: Bundle latest [JMX Metric Gatherer](https://github.com/open-telemetry/opentelemetry-java-contrib/tree/main/jmx-metrics) in installer packages and images for Windows and Linux ([#3262](https://github.com/signalfx/splunk-otel-collector/pull/3262))
```

rmfitzpatrick commented on tests/receivers/jmxreceiver/jmxreceiver_test.go:
```diff
@@ -0,0 +1,58 @@
+		"all_metrics_config.yaml", containers,
+		[]testutils.CollectorBuilder{
+			func(collector testutils.Collector) testutils.Collector {
+				// JMX requires a local directory that can be written to, so we must mount a local dir
+				// that the collector has write access to.
```
Doesn't this mean the jmx receiver won't work without this having been set in the collector image for end users? I think we should create an internal tmpdir that `TMPDIR` points to by default and document that this is required for `--tmpfs` if the collector is run w/ a read-only fs.

rmfitzpatrick commented on tests/receivers/jmxreceiver/testdata/all_metrics_config.yaml:
```diff
@@ -0,0 +1,27 @@
+processors:
+  resource:
+    attributes:
+      - key: telemetry.sdk.version
```
why is this deleted?

rmfitzpatrick commented on tests/receivers/jmxreceiver/jmxreceiver_test.go:
```diff
@@ -0,0 +1,58 @@
+
+				return collector.WithEnv(map[string]string{
+					"TMPDIR": tmp_dir,
+				}).WithMount(local_tmp, tmp_dir).WillFail(true)
```
This shouldn't fail

rmfitzpatrick commented:
Given what I'm seeing while updating some elements of this PR I think it'd be best to wait until 1.29.0 of the gatherer is released before including the jar with our artifacts (to gain https://github.com/open-telemetry/opentelemetry-java-contrib/commit/47f05dee61af2b912489932096dcaea12c054f15)

crobert-1 commented on tests/receivers/smartagent/telegraf-procstat/telegraf_procstat_test.go:
```diff
@@ -23,11 +23,7 @@ import (
-		expectedMetrics = "arm64.yaml"
-	}
-	testutils.AssertAllMetricsReceived(t, expectedMetrics, "all_metrics_config.yaml", nil, nil)
+	testutils.AssertAllMetricsReceived(t, "all.yaml", "all_metrics_config.yaml", nil, nil)
```
Is this a note for documenting behavior, or is this an action item to look into? Just want to make sure I understand here.

rmfitzpatrick replied:
This comment is describing the motivation for the included changes. It's not possible to use strace/ftrace against qemu tmk so tracking down may be tedious. I don't see much value in devoting more time to tracking down integration test issues that are likely caused by docker/binfmt/qemu when the underlying architecture has no conception of fs contents. Once we have actual arm runners then the distinctions are meaningful imo.

atoulme reviewed (approved):
LGTM.

jeffreyc-splunk commented on .gitlab/update-openjdk.sh:
```diff
@@ -31,6 +31,7 @@ create_collector_pr() {
     if [[ -n "$version" ]]; then
       echo ">>> Updating openjdk version to $version ..."
       sed -i "s|^ARG JDK_VERSION=.*|ARG JDK_VERSION=${version}|" internal/signalfx-agent/bundle/Dockerfile
+      sed -i "s|^ARG JDK_VERSION=.*|ARG JDK_VERSION=${version}|" cmd/otelcol/Dockerfile.windows
```
```suggestion
      sed -i "s|^ARG JDK_VERSION=.*|ARG JDK_VERSION=\"${version}\"|" cmd/otelcol/Dockerfile.windows
```
Alternatively, remove the quotes from `Dockerfile.windows` for the arg.  Otherwise, this script will open PRs just for the quotes.

jeffreyc-splunk commented on cmd/otelcol/Dockerfile:
```diff
@@ -35,6 +37,12 @@ RUN if [[ "$ARCH" = "amd64" || "$ARCH" = "arm64" ]]; then \
         mkdir -p /usr/lib/splunk-otel-collector/agent-bundle; \
     fi
 
+FROM alpine:3.17.0 AS jmx
```
```suggestion
FROM ${DOCKER_REPO}/alpine:3.17.0 AS jmx
```

jeffreyc-splunk commented on cmd/otelcol/Dockerfile:
```diff
@@ -45,14 +53,22 @@ FROM scratch
 ENV PATH=${PATH}:${SPLUNK_BUNDLE_DIR}/bin
+ENV PATH=${PATH}:${SPLUNK_BUNDLE_DIR}/jre/bin
+# avoid otel-java/autoconfigure default env vars for jmx receiver
+ENV OTEL_TRACES_EXPORTER=none
+ENV OTEL_LOGS_EXPORTER=none
```
Do we need to set these env vars for the deb/rpm packages?

rmfitzpatrick replied:
I think this should actually be set by the gatherer itself to avoid internal metrics (that aren't categorized) instead of adopting everywhere. Will update test to do this and open PR in that project.

jeffreyc-splunk commented on cmd/otelcol/Dockerfile.windows:
```diff
@@ -27,6 +27,25 @@ RUN Add-Type -AssemblyName System.IO.Compression.FileSystem ;\
+RUN setx /M PATH $(${Env:PATH} + \";${Env:JAVA_PATH}\")
+# update otel-java/autoconfigure defaults to prevent undesired telemetry
+RUN setx /M OTEL_TRACES_EXPORTER "none"
+RUN setx /M OTEL_LOGS_EXPORTER "none"
```
Should we add some basic tests to verify the paths and env vars to [CI](https://github.com/signalfx/splunk-otel-collector/blob/main/.github/workflows/win-package-test.yml#L362)?

rmfitzpatrick replied:
given https://github.com/signalfx/splunk-otel-collector/pull/3262#discussion_r1307944696 what would you suggest for the path check?

jeffreyc-splunk replied:
For now, maybe we can just try to execute something like `java -version` in the container?

jeffreyc-splunk reviewed (commented):
Should the [tar](https://github.com/signalfx/splunk-otel-collector/blob/main/internal/buildscripts/packaging/fpm/tar/build.sh) package also include the jar?

rmfitzpatrick commented:
> Should the [tar](https://github.com/signalfx/splunk-otel-collector/blob/main/internal/buildscripts/packaging/fpm/tar/build.sh) package also include the jar?

Yes, makes sense. @crobert-1 do you mind looking into this?

crobert-1 commented:
> Should the [tar](https://github.com/signalfx/splunk-otel-collector/blob/main/internal/buildscripts/packaging/fpm/tar/build.sh) package also include the jar?

I've added it to the TAR, and added a test to make sure it's there.

rmfitzpatrick commented on tests/receivers/jmxreceiver/jmxreceiver_test.go:
```diff
@@ -0,0 +1,50 @@
+)
+
+func TestJMXReceiverProvidesAllJVMMetrics(t *testing.T) {
+	// JMX metrics are missing when running on arm so the test only checks a subset
+	// of what would be received otherwise.
```
stale content

```suggestion
```

rmfitzpatrick commented on internal/buildscripts/packaging/msi/msi-builder/Dockerfile:
```diff
@@ -53,6 +53,7 @@ RUN cp /project/internal/buildscripts/packaging/msi/msi-builder/docker-entrypoin
     chmod a+x /docker-entrypoint.sh
 
 ENV OUTPUT_DIR=/project/dist
+ENV JMX_METRIC_GATHERER_RELEASE=1.28.0
```
```suggestion
ENV JMX_METRIC_GATHERER_RELEASE=1.29.0
```

rmfitzpatrick commented on internal/buildscripts/packaging/msi/msi-builder/build.sh:
```diff
@@ -26,6 +26,7 @@ FLUENTD_CONFD="/project/internal/buildscripts/packaging/msi/fluentd/conf.d"
 SUPPORT_BUNDLE_SCRIPT="/project/internal/buildscripts/packaging/msi/splunk-support-bundle.ps1"
 SPLUNK_ICON="/project/internal/buildscripts/packaging/msi/splunk.ico"
 OUTPUT_DIR="/project/dist"
+JMX_METRIC_GATHERER_RELEASE="1.28.0"
```
```suggestion
JMX_METRIC_GATHERER_RELEASE="1.29.0"
```

crobert-1 commented:
> Are any docs required? At a minimum, I think we should document that the packages/images will install/include the jar.

We currently don't have documentation for what's included in each package of the collector, but a jira issue has been opened to do the documentation work.

jeffreyc-splunk reviewed (approved):
LGTM. Please rebase and resolve conflicts.
